pkg/controllers/rollout/controller.go

```diff
@@ -0,0 +1,446 @@
+		return "", controller.NewExpectedBehaviorError(fmt.Errorf("cpr `%s` has no latest resourceSnapshot", crpName))
+	}
+	klog.V(2).InfoS("Find the latest associated resource resourceBinding", "clusterResourcePlacement", crpName,
+		"latestResourceSnapshotName", latestResourceSnapshotName)
+	return latestResourceSnapshotName, nil
+}
+
+// pickBindingsToRoll go through all bindings associated with a CRP and returns the bindings that are ready to be updated.
+// There could be cases that no bindings are ready to be updated because of the maxSurge/maxUnavailable constraints even if there are out of sync bindings.
+// Thus, it also returns a bool indicating whether there are out of sync bindings to be rolled to differentiate those two cases.
+func pickBindingsToRoll(allBindings []*fleetv1beta1.ClusterResourceBinding, latestResourceSnapshotName string, crp *fleetv1beta1.ClusterResourcePlacement) ([]*fleetv1beta1.ClusterResourceBinding, bool) {
+	// Those are the bindings that are chosen by the scheduler to be applied to selected clusters.
+	// They include the bindings that are already applied to the clusters and the bindings that are newly selected by the scheduler.
+	schedulerTargetedBinds := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// The content of those bindings that are considered to be already running on the targeted clusters.
+	readyBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that have the potential to be ready during the rolling phase.
+	// It includes all bindings that have been applied to the clusters and not deleted yet so that they can still be ready at any time.
+	canBeReadyBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that have the potential to be unavailable during the rolling phase which
+	// includes the bindings that are being deleted. It depends on work generator and member agent for the timing of the removal from the cluster.
+	canBeUnavailableBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that are candidates to be updated to be bound during the rolling phase.
+	boundingCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that are candidates to be removed during the rolling phase.
+	removeCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// Those are the bindings that are candidates to be updated to latest resources during the rolling phase.
+	updateCandidates := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// the list of bindings that are to be updated by this rolling phase
+	tobeUpdatedBinding := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+
+	// calculate the cutoff time for a binding to be applied before so that it can be considered ready
+	readyTimeCutOff := time.Now().Add(-time.Duration(*crp.Spec.Strategy.RollingUpdate.UnavailablePeriodSeconds) * time.Second)
```

Comment:
crp.Spec.Strategy is not a pointer.  RollingUpdate is. I am not 100% sure if it can be nil if there is a field inside that has a default value. I will test it